### ISSUE TITLE
Python 3.11 typing.Self not recognized

### ISSUE THREAD

**Issue by yanxke** [microsoft/pyright#4540]
**Describe the bug**
I have the latest version of VSCode/Pylance installed and it's giving me a red squiggle under the type hint Self.

**To Reproduce**
Pylance v2023.1.40, Python 3.11

**Comment by erictraut** [comment]
Please provide a self-contained minimal code sample.

**Comment by yanxke** [comment]
![image](https://user-images.githubusercontent.com/25883626/215223078-98584a56-5286-4504-af12-037a55a218ff.png)

**Comment by erictraut** [comment]
[PEP 673](https://peps.python.org/pep-0673/), which introduced the `Self` type, explicitly states that `Self` is not allowed in static methods. It even provides an example that shows that type checkers should reject the use of `Self` in this context. Here's the rationale from the PEP:

> Note that we reject Self in staticmethods. Self does not add much value since there is no self or cls to return. The only possible use cases would be to return a parameter itself or some element from a container passed in as a parameter. These don’t seem worth the additional complexity.

So, this behavior is correct and is mandated by the PEP.

**Comment by yanxke** [comment]
Thanks for the explanation, though the specification is unintuitive.  I was simply trying to specify the class object itself as the return type.  I suppose I still have to use `from __future__ import annotations` for now.

**Comment by alex-ball** [comment]
@yanxke It depends on whether the factory should still return a `Foo` even on subclasses of Foo. If so:

```py
class Foo():
    @staticmethod
    def factory() -> "Foo":
        return Foo()
```

But if it should always return an instance of the class whose factory you're calling (which is more likely what you want) then this works:

```py
class Foo():
    @classmethod
    def factory(cls) -> Self:
        return cls()
```
